app/src/test/java/org/mozilla/tv/firefox/webrender/cursor/CursorViewModelTest.kt

```diff
@@ -0,0 +1,178 @@
+            isVoiceViewEnabled = MutableLiveData()
+            `when`(it.isVoiceViewEnabled).thenReturn(isVoiceViewEnabled)
+        }
+
+        screenController = mock(ScreenController::class.java).also {
+            activeScreen = MutableLiveData()
+            `when`(it.currentActiveScreen).thenReturn(activeScreen)
+        }
+
+        sessionRepo = mock(SessionRepo::class.java).also {
+            sessionState = MutableLiveData()
+            `when`(it.state).thenReturn(sessionState)
+        }
+
+        viewModel = CursorViewModel(frameworkRepo, screenController, sessionRepo)
+    }
+
+    // This method pattern is duplicated but it's significantly more readable this way.
+    @Test
+    fun `GIVEN VoiceView is enabled THEN the cursor is disabled`() {
+        val (initialScreen, restScreens) = firstAndRestScreens()
+        val (initialUrl, restUrls) = POSSIBLE_URLS.firstAndRest()
+
+        // 1 event is emitted across all initial values: no events are emitted until all initial values are set.
+        val emittedEventCount = 1 + restScreens.size + restUrls.size
+
+        viewModel.isEnabled.assertValueForEmissionCount(false, emittedEventCount) {
+            isVoiceViewEnabled.value = true
+
+            activeScreen.value = initialScreen
+            sessionState.value = newStateWithUrl(initialUrl)
+
+            // Now that initial value is set, iterate over remaining states. Ideally this would be all permutations.
+            restScreens.forEach { activeScreen.value = it }
+            restUrls.forEach { sessionState.value = newStateWithUrl(it) }
+        }
+    }
+
+    @Test
+    fun `GIVEN url is youtube TV THEN cursor is disabled`() {
+        val (initialScreen, restScreens) = firstAndRestScreens()
+
+        // 1 event is emitted across all initial values: no events are emitted until all initial values are set.
+        val emittedEventCount = 1 + 1 /* isVoiceViewEnabled bool */ + restScreens.size
+
+        viewModel.isEnabled.assertValueForEmissionCount(false, emittedEventCount) {
+            sessionState.value = newStateWithUrl("https://youtube.com/tv/")
+
+            activeScreen.value = initialScreen
+            isVoiceViewEnabled.value = false
+
+            // Now that initial value is set, iterate over remaining states. Ideally this would be all permutations.
+            restScreens.forEach { activeScreen.value = it }
+            isVoiceViewEnabled.value = true
+        }
+    }
+
+    @Test
+    fun `GIVEN web render fragment is not visible THEN cursor is disabled`() {
+        val (initialScreen, restScreens) = ActiveScreen.values()
+            .filter { it != WEB_RENDER } // web render is not visible.
+            .firstAndRest()
+        val (initialUrl, restUrls) = firstAndRestUrls()
+
+        // 1 event is emitted across all initial values: no events are emitted until all initial values are set.
+        val emittedEventCount = 1 + 1 /* isVoiceViewEnabled bool */ + restUrls.size + restScreens.size
```

Comment:
Especially ones that duplicate said comments in every test! 😱 

One way would be if the test could generate all the permutations of the test data, it could also count the expected number of event emissions. Then it can assert the emission count with the given value. Something like:
```kotlin
isEnabled.assertEmissionsOverAllPermutations(expected = false, inputData = [...])
```

But that's hard.

It seems really unsatisfactory that it's non-trivial to generate and test all the permutations of test data: @Baron-Severin does RxJava have test libraries that can do this for you?

Reply:
Not that I'm aware of.  There are Java libraries for generating permutations that we could pull in as test dependencies, and then set up our tests similarly to imitation JUnit 5 parameterized tests.  If we're going to put this work in, though, it probably 1) need to be a separate story, and 2) be general enough to cover use cases beyond a series of booleans.  That's difficult.

FWIW, I know there is some research into the area of generating test cases more generally, but I've personally only heard about it in a very functional programming space, and (I believe) it's somewhat nascent even there.